jurity/recommenders/ctr.py

```diff
@@ -128,22 +183,94 @@ def get_score(self, actual_results: pd.DataFrame, predicted_results: pd.DataFram
 
     def _get_extended_results(self, results: List[np.ndarray]) -> dict:
         results = np.concatenate(results)
         return {'ctr': np.mean(results), 'support': results.size}
 
+    def _get_matching_ctr(self, actual_results, sorted_clicks, batch_accumulate, return_extended_results):
+        matches = self._get_matches(actual_results, sorted_clicks)
+        clicks = matches[self.value_column].values
+
+        return self._accumulate_and_return(clicks, batch_accumulate, return_extended_results)
+
+    def _get_ips(self, actual_results, sorted_clicks, n_items, batch_accumulate, return_extended_results):
+        matches, actual_results = self._get_match_probabilities(actual_results, sorted_clicks, n_items)
+
+        # calculate reward / propensity for the matches
+        matches[Constants.inverse_propensity] = matches[self.value_column] / matches[self.propensity_column]
+
+        # final calculation requires the full n of the data set
+        # merge with sorted_clicks and impute with 0 to include the non-matches
+        sorted_clicks = sorted_clicks.merge(matches[[Constants.inverse_propensity]],
+                                            how='left', left_index=True, right_index=True)
+        sorted_clicks[Constants.inverse_propensity].fillna(0, inplace=True)
+
+        ips = sorted_clicks[Constants.inverse_propensity].values
+
+        return self._accumulate_and_return(ips, batch_accumulate, return_extended_results)
+
+    def _get_doubly_robust_estimate(self, actual_results, sorted_clicks, n_items, batch_accumulate,
+                                    return_extended_results):
+        matches, actual_results = self._get_match_probabilities(actual_results, sorted_clicks, n_items)
+
+        # calculate (actual - predicted) / propensity for the matches
+        matches[Constants.ips_correction] = (matches[self.value_column] - matches[self.value_column+'_r']) \
+                                           / matches[self.propensity_column]
+
+        # merge with sorted_clicks and impute correction with 0 to include the non-matches for DR calculation
+        sorted_clicks = sorted_clicks.merge(matches[[Constants.ips_correction]],
+                                            how='left', left_index=True, right_index=True)
+        sorted_clicks[Constants.ips_correction].fillna(0, inplace=True)
+
+        # DR = predicted + propensity correction
+        sorted_clicks[Constants.estimate] = sorted_clicks[self.value_column] + sorted_clicks[Constants.ips_correction]
+
+        dr = sorted_clicks[Constants.estimate].values
+
+        return self._accumulate_and_return(dr, batch_accumulate, return_extended_results)
+
+    def _get_match_probabilities(self, actual_results, sorted_clicks, n_items):
+
+        if self.propensity_column not in actual_results.columns:
+
+            actual_results[self.propensity_column] = 1/n_items
```

Comment:
I assume we would not have 0-division issue, if we make it until here

Reply:
n_items is the length of the list of unique items in the item id column. Even if it is all NaNs, that's one item in the list so guaranteed to be at least 1.

Reply:
cool!